highfidelity/hifi
# Case 21457, Case 16836, Case 15782: Improve AudioInjector threading

https://highfidelity.manuscript.com/f/cases/21457/CRASH-while-using-Appreciate-App
https://highfidelity.manuscript.com/f/cases/16836/Remove-AudioInjector-stopInjectorImmediately-from-the-API
https://highfidelity.manuscript.com/f/cases/15782/AudioInjector-getOptions-function-returns-undefined

Test plan:
- You'll need at least two people running this build to test non-local sounds.
- A new stat has been added in the third column of the expanded stats: "Injectors (Local/NonLocal)".  As you make sounds during testing, these numbers should reflect the sounds you have playing.  For example, if you play a non-looping local sound, the first number will go up by 1 and then go down by 1 when the sound finishes.  If you play a non-local sound, both numbers will go up.
- The keyboard sounds, snapshot sound, various tablet sound effects, and avatar collision sounds should work as in master.  The first three will be local only, the latter should be heard by others, I believe.
- Text to speech should still work.
- Stress test the appreciate app.  The other person should be able to hear your clapping/whistling, and neither of you should crash.
- Run [this](https://gist.githubusercontent.com/SamGondelman/9c51bbe0894ae3cabc7d729bfa721976/raw/04edd754c301478ce05f8c322eb5e32847f3afd1/AudioInjectorTest.js).  Stand near the other person.
- Press "y".  A local sound will play.  It will repeat when it finishes.  Press "y" a lot and it'll repeatedly restart.
- Press "u" once.  The other person should be able to hear it now.  Press "y" a lot and it'll repeatedly restart for both of you.
- Press "o" once.  The sound will play for both of you and then stop once it's finished.  Your injector count in the stats will go to 0.  Press "y" a lot and it'll repeatedly restart for both of you.
- Press "o" again so that the sound loops now.  Press "h" and 4 lines will print in the logs, something like:
```
[03/13 12:53:19] [DEBUG] [hifi.scriptengine.script] [SoundTest.js] boop
[03/13 12:53:19] [DEBUG] [hifi.scriptengine.script] [SoundTest.js] true
[03/13 12:53:19] [DEBUG] [hifi.scriptengine.script] [SoundTest.js] 0.027085280045866966
[03/13 12:53:19] [DEBUG] [hifi.scriptengine.script] [SoundTest.js] {"position":{"x":7.869675755500793e-8,"y":-10.463151931762695,"z":-2.9709190130233765e-7},"volume":1,"loop":true,"orientation":{"x":0,"y":0,"z":0,"w":1},"ignorePenumbra":false,"localOnly":false,"secondOffset":0,"pitch":1}
```
- While the sound is playing, stop the script.  The sound should stop immediately for both of you.
